hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -2042,30 +2043,66 @@ private static ClosableIterator<HoodieRecord> createSecondaryIndexGenerator(Hood
         Option.empty());
     ClosableIterator<HoodieRecord> fileSliceIterator = ClosableIterator.wrap(fileSliceReader);
     return new ClosableIterator<HoodieRecord>() {
+      private HoodieRecord nextValidRecord;
+
       @Override
       public void close() {
         fileSliceIterator.close();
       }
 
       @Override
       public boolean hasNext() {
-        return fileSliceIterator.hasNext();
+        // As part of hasNext() we try to find the valid non-delete record that has a secondary key.
+        if (nextValidRecord != null) {
+          return true;
+        }
+
+        // Secondary key is null when there is a delete record, and we only have the record key.
+        // This can happen when the record is deleted in the log file.
+        // In this case, we need not index the record because for the given record key,
+        // we have already prepared the delete record before reaching this point.
+        // NOTE: Delete record should not happen when initializing the secondary index i.e. when called from readSecondaryKeysFromFileSlices,
+        // because from that call, we get the merged records as of some committed instant. So, delete records must have been filtered out.
+        // Loop to find the next valid record or exhaust the iterator.
+        while (fileSliceIterator.hasNext()) {
```

Comment:
are you sure we need this logic?
So, in general, a fileSliceReader can return deleted record too is it? whats the purpose? 
we generally don't return deleted records from say LogRecordReader right. 
so, why a fileSliceReader might return a deleted record. 
can you help me understand.